### PR TITLE
Add Generation Pipeline Documentation

### PR DESCRIPTION
**Pull request by kwridan**
### Short description 📝

The generation pipeline has undergone numerous changes and has evolved over the past year or so, it now has several steps and processes which may be hard to follow or explain.

### Solution 📦

To help onboard new contributors as well as allow us to evaluate where certain features can be implemented some additional documentation is needed for the generation pipeline.

### Implementation 👩‍💻👨‍💻

- [x] Add generation pipeline contributors documentation
- [x] Include the source xml for the draw.io digram to allow future modifications to the diagrams

### Notes 📝

- We can slowly add other similar pages going forward for the different processes, e.g. Manifest Loading, Signing, Caching etc...

### CODE REVIEW

File: website/markdown/docs/contribution/generation-pipeline.mdx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+- **Side Effects Descriptors**: Those are descriptions of side effects Tuist needs to perform
+    - Those are quite low level and basic (e.g. create a directory, file, etc...)
+- **Workspace/Project Descriptors**: Those are in-memory representations of the low level `XcodeProj` types
+- **Interactors**: Those are components that perform specialized side effects at the end of the generation process (e.g. signing / SwiftPM resolution etc...)
+
+## Pipeline
+
+The pipeline consists of the following stages & steps that are performed sequentially. The main orchestration of all of these steps resides within `TuistKit.Generator`.
+
+![Generation Pipeline Overview](assets/generation-pipeline/overview.png)
```
Comment (kwridan) [inline website/markdown/docs/contribution/generation-pipeline.mdx:40]: I was struggling to name the different logical groups, suggestions welcome :)

File: website/markdown/docs/contribution/generation-pipeline.mdx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+- **Manifest**: Those are the public user facing files used to describe projects. (e.g. `Project.swift`, `Workspace.swift`, `Config.swift`, etc...)
+    - The types are declared in the `ProjectDescription` target
+    - Those types can't depend on any other targets/modules as the `ProjectDescription` target is shipped as a standalone dynamic library with each Tuist release
+- **Model**: Those are the internal model types used to represent various entities of the Tuist Graph and supporting types.
+    - The types are declared in the `TuistCore` target
+    - Many types are similar to the ones defined in the `ProjectDescription` target
+        - Model types however contain more information than their `ProjectDescription` counterpart (e.g. expanded paths)
+        - Model types are often optimized for simpler internal usage
+- **Graph**: This is a collection of Tuist **Model**s along with all their loaded dependency information in a graph representation
+    - This is the primary entity used to at the heart of many Tuist features including project generation
```
Comment (fortmarek) [inline website/markdown/docs/contribution/generation-pipeline.mdx:23]: > used to at

I think `to` can be dropped here 🙂

File: website/markdown/docs/contribution/generation-pipeline.mdx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+- **Model Mapper**: This is a type used to transform individual **Model**s ahead of loading the graph
+    - There 3 different sub-types `WorkspaceMapper`, `ProjectMapper` and `TargetMapper` which can be used to transform the corresponding model type
```
Comment (fortmarek) [inline website/markdown/docs/contribution/generation-pipeline.mdx:25]: > There 3

You probably meant "There are 3 ... "?

File: website/markdown/docs/contribution/generation-pipeline.mdx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    - In addition to transforming models, mappers can also declare a set of **Side Effects Descriptors** to perform at the end of the generation pipeline
+    - The mappers themselves do not perform the side effects
+- **Graph Mapper**: This is a type used to transform the entire **Graph**
```
Comment (fortmarek) [inline website/markdown/docs/contribution/generation-pipeline.mdx:28]: maybe this can be added to `Model Mapper` section? The subcomments for this are almost identical.

File: website/markdown/docs/contribution/generation-pipeline.mdx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+![Generation Pipeline Overview](assets/generation-pipeline/overview.png)
+
+### Loading Manifests & Models
+
+![Manifest & Model Loading](assets/generation-pipeline/manifest-model-loading.png)
+
+- **Load Manifests**: The (`ProjectDescription` types) are loaded from disk using `ManifestLoader`
+    - This involves performing `swiftc` calls to compile and parse the contents of the manifests
+    - Additionally if `ProjectDescriptionHelpers` are used, those are also complied and cached
```
Comment (fortmarek) [inline website/markdown/docs/contribution/generation-pipeline.mdx:48]: `complied` > `compiled`

File: website/markdown/docs/contribution/generation-pipeline.mdx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+- **Lint Manifests**: Manifests are undergo a series of checks that may produce warnings or fail the generation process
```
Comment (fortmarek) [inline website/markdown/docs/contribution/generation-pipeline.mdx:49]: > Manifests are undergo

I think `undergo` can be dropped?

File: website/markdown/docs/contribution/generation-pipeline.mdx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+     - Manifests are converted concurrently
+
+### Loading the Graph
+
+![Graph Loading](assets/generation-pipeline/graph-loading.png)
+
+- **Apply Model Mappers**: Models undergo a few additional transformations via model mappers which modify the models and produce `SideEffectDescriptors`
+    - The model mappers are orchestrated in `WorkspaceMapperProvider` and `ProjectMapperProvider`
+- **Load the Graph**: The transformed Models are used to load the Graph using `GraphLoader`
+- **Apply Graph Mappers**: The Graph undergos a few additional transformations via graph mappers which modify the graph and produce an additional set of `SideEffectDescriptors`
```
Comment (fortmarek) [inline website/markdown/docs/contribution/generation-pipeline.mdx:62]: `undergos` -> `undergoes`

**Review by fortmarek — APPROVED** [review]
Thanks for doing this, this will be really handy for me and other contributors in the future! Just smaller nitpicks, but I like the overall text + the images are super helpful 👏

File: website/markdown/docs/contribution/generation-pipeline.mdx (hunk 4 above)
Comment (kwridan) [reply website/markdown/docs/contribution/generation-pipeline.mdx]: They are indeed similar, the graph mappers are a separate set of components though hence why it's called out separately. I've abbreviated it to reduce the repetition 👍

Comment (fortmarek) [reply website/markdown/docs/contribution/generation-pipeline.mdx]: looks good, thanks!

**Comment by kwridan** [comment]
Thanks @fortmarek, It's always funny reading back what I thought I had reviewed for typos/errors 😅 and discovering there's quite a few left!

**Comment by pepibumur** [comment]
Wow, thanks a lot @kwridan for writing this up. It'll be a very useful reference for new contributors joining the project.
